File: apps/api/internal/models/case_study.go
package models

import (
	"time"

	"gorm.io/datatypes"
	"gorm.io/gorm"

	"megagig-software-solution/apps/api/internal/ids"
)

// CaseStudy represents a casestudy in the system.
type CaseStudy struct {
	ID            string                      `gorm:"primarykey;size:36" json:"id"`
	Slug          string                      `gorm:"size:255" json:"slug" binding:"required"`
	ClientName    string                      `gorm:"size:255" json:"client_name" binding:"required"`
	Tagline       string                      `gorm:"size:255" json:"tagline" binding:"required"`
	CategoryTags  datatypes.JSONSlice[string] `gorm:"type:json" json:"category_tags"`
	StatusBadge   string                      `gorm:"size:255" json:"status_badge" binding:"required"`
	HeroImageID   string                      `gorm:"size:36;index" json:"hero_image_id" binding:"required"`
	HeroImage     Upload                      `gorm:"foreignKey:HeroImageID" json:"hero_image"`
	Problem       string                      `gorm:"type:text" json:"problem"`
	WhatWeBuilt   string                      `gorm:"type:text" json:"what_we_built"`
	Result        string                      `gorm:"type:text" json:"result"`
	TechStack     datatypes.JSONSlice[string] `gorm:"type:json" json:"tech_stack"`
	LiveURL       string                      `gorm:"size:500" json:"live_url"`
	TestimonialID string                      `gorm:"size:36;index" json:"testimonial_id"`
	Testimonial   *Testimonial                `gorm:"foreignKey:TestimonialID" json:"testimonial,omitempty"`
	Published     bool                        `json:"published"`
	SortOrder     int                         `json:"sort_order"`
	Version       int                         `gorm:"not null;default:1" json:"version"`
	CreatedAt     time.Time                   `json:"created_at"`
	UpdatedAt     time.Time                   `json:"updated_at"`
	DeletedAt     gorm.DeletedAt              `gorm:"index" json:"-"`
}

// BeforeCreate generates a UUID before inserting.
func (m *CaseStudy) BeforeCreate(tx *gorm.DB) error {
	if m.ID == "" {
		m.ID = ids.New()
	}
	return nil
}

// BeforeUpdate increments Version so offline clients can detect server-side updates.
func (m *CaseStudy) BeforeUpdate(tx *gorm.DB) error {
	tx.Statement.SetColumn("version", gorm.Expr("version + 1"))
	return nil
}

File: packages/shared/schemas/case-study.ts
import { z } from "zod";

export const CreateCaseStudySchema = z.object({
  slug: z.string().min(1, "Required"),
  client_name: z.string().min(1, "Required"),
  tagline: z.string().min(1, "Required"),
  category_tags: z.array(z.string()).optional(),
  status_badge: z.string().min(1, "Required"),
  hero_image_id: z.string().uuid("Invalid ID"),
  problem: z.string().optional(),
  what_we_built: z.string().optional(),
  result: z.string().optional(),
  tech_stack: z.array(z.string()).optional(),
  live_url: z.string().optional(),
  testimonial_id: z.string().uuid("Invalid ID").optional(),
  published: z.boolean().optional(),
  sort_order: z.number().int().optional(),
});

export const UpdateCaseStudySchema = z.object({
  slug: z.string().min(1, "Required").optional(),
  client_name: z.string().min(1, "Required").optional(),
  tagline: z.string().min(1, "Required").optional(),
  category_tags: z.array(z.string()).optional(),
  status_badge: z.string().min(1, "Required").optional(),
  hero_image_id: z.string().uuid("Invalid ID").optional(),
  problem: z.string().optional(),
  what_we_built: z.string().optional(),
  result: z.string().optional(),
  tech_stack: z.array(z.string()).optional(),
  live_url: z.string().optional(),
  testimonial_id: z.string().uuid("Invalid ID").optional(),
  published: z.boolean().optional(),
  sort_order: z.number().int().optional(),
});

export type CreateCaseStudyInput = z.infer<typeof CreateCaseStudySchema>;
export type UpdateCaseStudyInput = z.infer<typeof UpdateCaseStudySchema>;
